Memory cleanup at exit

dustanddreams opened the pull request:
This is a ~work-in-progress~ support of "memory cleanup at exit" code (`OCAMLRUNPARAM=c`), see issue #10865.

This is implemented in two parts:
1. perform the equivalent of what `domain_terminate` does in the main thread upon exit.
2. invoke the finalizer, if any, for all remaining objects in the shared heap.

This appears to work when only one domain is left running at program termination time. When multiple domains are used, no effort to release memory is achieved due to the domains being in an unknown state. However, if memory cleanup at exit mode is enabled, the domain threads will be forcibly stopped.

ghost commented:
~In order to make progress for what is probably most of the use cases of that feature, I've currently restricted the cleanup to single-domain operation.~

~In this state, there are no segmentation faults or other misbehaviours upon exit, only memory leaks (i.e. it's as bad as before in the worst case.)~

jmid commented:
Allow me summarize my understanding of this work (feel free to correct me):
- with this PR both single-domain programs and "well-behaved multi-domain programs"[^1] will free the malloc'ed memory at exit
- this will only happen in  "memory cleanup at exit" mode (`OCAMLRUNPARAM=c`)

This strikes me as a good first step to get an initial "memory cleanup at exit" mode working in OCaml 5.

[^1]: multi-domain programs that `Domain.join` any child `Domain`s before reaching program exit, meaning that they will be single-domain at that point

ghost commented:
Your understanding is correct.

NickBarnes commented:
Possibly I am misunderstanding something here. In the current runtime, when a domain terminates, its major heap (a.k.a. "shared heap") is "orphaned" by `caml_teardown_shared_heap`, and stashed away in the (misleadingly-named) `pool_freelist` until the next GC "cycle" when it is adopted by some other domain (in `caml_cycle_heap`). The memory is not freed, nor can it be as there may be pointers into it from the major heaps of other domains. Likewise, finalisers are not run because the blocks may still be alive. This change appears to break that, by freeing the shared heap of a domain in `domain_terminate` (via `caml_teardown_domain` calling `caml_teardown_shared_heap` which calls `caml_finalise_heap`, etc).
I think that, instead, we need to call `caml_finalise_heap` (which should do approximately what it does in this PR, i.e. iterate through the pool lists and large alloc linked list) _only when the last domain terminates_. When some other domain terminates, its heap should continue to be orphaned and then adopted, as at present; blocks in that heap which die due to the domain termination will be collected (and finalisers run) in the normal way at a subsequent collection.

gasche commented:
(This PR is related to #13010 and I was not sure which one to comment in.)

Here is the behavior I would expect for `caml_shutdown`:
1. all mutator threads are stopped
2. all domains are terminated
3. all resources allocated by the runtime are released

In 4.x, the cleanup-at-exit mode would make `caml_do_exit` call `caml_shutdown`, and I see no harm in keeping the same behavior in 5.x. (There the use-case is not an OCaml plugin within a larger non-OCaml program, as we know the process will exit anyway, it is to use `valgrind` to detect resources leaks in our C code or something like this.) If we are not in cleanup-at-exit mode, we don't need to do anything.

I could see a use-case for another function that does a "soft shutdown" by raising the `Thread.Exit` exception in all mutator threads, and letting the mutator themselves cleanup their own resources. This is nice, it corresponds to @damiendoligez's suggestion, it is more work and could be dealt with separately.

ghost commented:
This new version incorporates @NickBarnes's feedback, but should not be considered ready yet. There are still issues when using the debug runtime. On the other hand it will be easier to review as the cleanup code paths are more visible now.

ghost commented:
Status update:
- single-domain programs, or those which `Domain.join` all the running domains prior to exiting, ought to have all their memory allocations released correctly.
- if there are "rogue" domains still running, a message is `caml_gc_log`ed, and an attempt will be made to cause them to stop, but no particular attempt will be made to clean up their resources (this is considered too risky). If these domains are stuck in non-caml code (e.g. waiting for a resource in one of the `pthread` functions), then they won't get interrupted, and no attempt is currently made to forcibly `pthread_kill` them.

gasche commented on runtime/domain.c:
```diff
@@ -2096,6 +2127,105 @@ static void domain_terminate (void)
+  int i = atomic_load_acquire(&caml_num_domains_running) + 5;
+  while (!caml_domain_alone() && i != 0) {
+#ifdef WIN32
+    Sleep(1);
```
This sleep-based synchronization looks very fishy to me -- but then I'm not a domain expert. Is there not a standard synchronization primitive that we can use?

dustanddreams replied:
It's only ugly (-:

If we reach this state, there is absolutely no guarantee that the interrupt request will be processed by the domains, so we can not wait for them without an upper time bound, in case at least one domain is unresponsive. The time limit here is (number of currently running domains + 5) x 1 millisecond. That's an arbitrary number which hopefully allows every non-sleeping domain to be scheduled and receive the exit request.

gasche replied:
I understand the concern, but I am not sure what scenarios you have in mind about delayed processing.

I am assuming that all domains have at least a thread that checks for interruptions on a regular basis -- otherwise this is a bug that should be fixed to restore this property, because the minor GC relies on it. I am not sure if you make a different assumption, or if you also assume this but then are thinking of the *other* threads on that same domain that are waiting on other blocking calls, or something else?

As long as the owner of the domain lock regularly handles interruptions, we know (I think?) that the other threads on the domain will not use the runtime again before they take the domain lock. So we could have a check for "we are forced to terminate" at that point that would shut them down before they return to OCaml code, and not have to wait at all for those threads right now.

gasche commented:
The implementation mechanism for interrupting all domains in this PR is somewhat custom: it uses the same interruption mechanism as STW sections, but it is not implemented as a STW itself, instead it is a new code path in `handle_incoming`.

I find it hard to reason about, and in particular there are two subtleties that may or may not be handled properly:

1. What happens if a new domain is being created while we are trying to shutdown? You added a test for this in `caml_domain_spawn`, but I don't know if the test is enough. (What happens if a spawn was ongoing, exactly after the test, when you call shutdown?). In contrast, the STW mechanism has careful synchronization in place to block new spawns while a STW section is running, so reusing it may make things more correct here.

2. What happens if `caml_shutdown` is called while some threads are blocked, and the backup thread handles the shutdown interruption? I suppose that the backup thread may cleanup the domain structure, but then what happens when another thread on the domain is done with their blocking call and decides to take the domain lock and run OCaml code again?

I think that a robust solution should come with an explanation of its interaction with the other domain lifetime events (see for example the existing documentation comments about the backup thread cycle above `BT_IN_BLOCKING_SECTION`, and the synchronization mechanisms of STW section above `caml_try_run_on_all_domains_with_spin_work`), and of how it preserves or modifies the invariants governing several threads morally attached to the same domain (including the backup thread).

If I was trying to implement this, I would try the following:
- Implement the shutdown/cancellation logic inside the STW mechanism, possibly by just calling a STW section whose body calls `pthread_exit` for most domains, but maybe this does not work (it break the logic that comes in `stw_handler` after the callback) and the logic has to be added to `stw_handler` or some other part of the STW mechanism, conditioned on the global "we are forced to terminate" condition.
- Check the "we are forced to terminate" condition in `caml_leave_blocking_section` (or maybe `caml_bt_enter_ocaml`, but I think that this may not work in Thread-using program), think about which other places need such a check, and try to find a crisp explanation for which places need the check and why.

ghost commented:
> The implementation mechanism for interrupting all domains in this PR is somewhat custom: it uses the same interruption mechanism as STW sections, but it is not implemented as a STW itself, instead it is a new code path in `handle_incoming`.

An earlier work (#13010) did use an STW for convenience, and some people commented this was probably overkill. But I agree that it was simpler to write as an STW rendezvous.

> I find it hard to reason about, and in particular there are two subtleties that may or may not be handled properly:
> 
>     1. What happens if a new domain is being created while we are trying to shutdown? You added a test for this in `caml_domain_spawn`, but I don't know if the test is enough. (What happens if a spawn was ongoing, exactly after the test, when you call shutdown?). In contrast, the STW mechanism has careful synchronization in place to block new spawns while a STW section is running, so reusing it may make things more correct here.

If a spawn was ongoing and had already passed the added test, then it will spawn correctly, and then honour the exit request (almost) immediately.

>     2. What happens if `caml_shutdown` is called while some threads are blocked, and the backup thread handles the shutdown interruption? I suppose that the backup thread may cleanup the domain structure, but then what happens when another thread on the domain is done with their blocking call and decides to take the domain lock and run OCaml code again?

Argh, this is something I did not take into account. The interrupt handling needs to ignore the exit request if processed in the backup thread, i.e. if `caml_bt_is_self`.

gasche commented:
> An earlier work (https://github.com/ocaml/ocaml/pull/13010) did use an STW for convenience, and some people commented this was probably overkill. But I agree that it was simpler to write as an STW rendezvous.

Ah, I suppose you had comment https://github.com/ocaml/ocaml/pull/13010#issuecomment-2009204051 in mind. It is possible that I am contradicting myself from 2 months ago, but I think that (1) notifying other domains that they should terminate (as we discussed then) should not require a STW (and I don't think that raising an exception from within a STW is currently supported), but that (2) cleaning up the runtime state shared by all domains (as discussed now) requires a synchronization with all mutators, which could be done with a STW. In any case, "simpler" is good here, as long as it aligns with "reasonably easy to convince ourselves that it is correct".

gasche commented:
Aside: I thought of trying to implement a domain-safe `caml_shutdown` myself to try to understand this better, and then I remembered that writing tests for this is tedious. I thought of reusing your own testsuite, but I don't see any new tests in the current draft PR. How do you do your testing?

ghost commented:
> It is possible that I am contradicting myself from 2 months ago

You are not contradicting yourself - your opinion on the problem evolves as you are getting more familiar with it.

ghost commented:
> Aside: I thought of trying to implement a domain-safe `caml_shutdown` myself to try to understand this better, and then I remembered that writing tests for this is tedious. I thought of reusing your own testsuite, but I don't see any new tests in the current draft PR. How do you do your testing?

There are several tests in the existing testsuite which, in their current state, exit with multiple domains running, and which are representative enough:
- ~`tests/lazy/lazy3.ml`~
- `tests/parallel/backup_thread.ml`
- `tests/parallel/major_gc_wait_backup.ml`
- `tests/parallel/mctest.ml`

kayceesrk commented:
The PR has been marked as ready for review. @dustanddreams, can you summarise what has improved since it was set to draft a few weeks ago (https://github.com/ocaml/ocaml/pull/12964#event-13067669160)?

Is the original PR message still an accurate summary of the guarantees provided? https://github.com/ocaml/ocaml/pull/12964#issue-2122375997.

ghost commented:
I have updated the PR description to match the current reality.

Short summary: when cleanup mode is enabled, memory is released only if only one domain is left running at program shutdown, otherwise the leftover threads are cancelled but nothing else is done for we can't assume anything about the state they are in. When cleanup mode is not enabled, there is no change of behaviour at all.

ghost commented:
Note that changes are to be expected after #13272 goes in, since there will be more dynamic allocations to clean up. If you're short on time, I'd suggesting postponing your review until then.

gasche commented:
I don't think I will be able to do a full review soon, but I had a quick look at it seems that the PR is moving in the direction of being simpler, thanks! On the other hand, I remain unconvinced by the reasoning in `stw_terminate_domain`, whose documentation explains that "we are not in a state where we can safely release resources", and that therefore no resource cleanup is attempted.

Here is my intuition about this -- but it may very well be incorrect/simplistic.

- The thread on each domain that answers the STW interrupt owns the domain resources.
- So in fact it could release the domain resources right on the spot.
- We could either kill the other threads running on the domain at this point, or be more lazy by just making sure that they stop themselves if they regain the domain lock after termination has been ordered (before they try to access domain resources), or both.

If I was writing this myself, I would try to ensure that when a non-backup thread regains the domain lock, it starts by checking whether the domain has already been shut down, and terminating itself in this case. (Can we check this just by looking at `domains_exiting`? We want to be sure that the STW section has run to release all resources on the domain before we switch to this behavior, so maybe we need a separate boolean `domains_have_exited`, or we need to set `domains_exiting` at the *end* of the `stw_terminate_domain` STW section, not before it as currently.)

gasche commented on runtime/domain.c:
```diff
@@ -2096,6 +2115,111 @@ static void domain_terminate (void)
+  atomic_store_relaxed(&domains_exiting, 1);
+
+  pthread_t myself = pthread_self();
+  caml_try_run_on_all_domains(stw_terminate_domain, &myself, NULL);
```
We probably want to run this in a loop, because `_try_` indicates that it may do nothing and return `false` if it runs into a concurrent STW request from another domain -- see other use-cases, one may either loop until the call return `1` or check another variable that indicates that this domain or some other domain succeeded in starting a STW section that achieved the desired purpose.

dustanddreams replied:
You're right. In a former version, this used to be part of a `while (!caml_domain_alone())` loop.

Commit:
Rework backup thread termination.

Make the value of the backup_thread_running per-domain field reliable.

ghost commented:
Status update:
- the current code seems to be reliable enough, I've split it into multiple parts to help review.
- when compiling with `CC=gcc -fsanitize=leak` and `OCAMLRUNPARAM=c=1,v=0`, the only test failures (13) are either tests which do not work with the sanitizer (`tests/backtrace/pr2195.ml`), have memory leaks in non-OCaml code (`tests/embedded/cmcaml.ml`), or do not end gracefully (i.e. there are threads or domains still running), in which case the cleanup is only partial.
- when compiling with tsan, there are false race reports between domain threads and the deallocation of the `all_domains` array; I have not figured out yet a good way to silence tsan here (which is why the PR is still marked as draft).
- I can't reproduce the CI failure for `testsuite/tests/parallel/backup_thread.ml` in bytecode.

Commit:
Add infrastructure to force all domain threads to exit.

Commit:
Bring back "memory cleanup at exit" mode, step 1.

Upon shutdown, if memory cleanup is requested, make sure all domain and
backup threads are terminated, and invoke the domain termination function
for the last running domain if it was alone at shutdown time.

Commit:
Bring back "memory cleanup at exit" mode, step 2.

Upon shutdown, if memory cleanup is requested, release shared heap
memory.

Commit:
Fix memory leak on unlikely error.

Commit:
Document the return of memory cleanup mode.

Commit:
Skip freeing some memory during cleanup if TSan is enabled.

Commit:
Correctly account for mutex locks after forks.

ghost commented:
Sorry, there had been a slight mistake in the first commit ("Rework backup thread termination.") which was exposed by the CI when running the testsuite in parallel.
This is now fixed in the commits pushed earlier today.

gasche commented:
@dustanddreams I can see that you are rebasing this PR regularly, but I am a bit fuzzy about its current status.

- Is it ready for review? My understanding is that you think it is, but are maybe to shy/polite to explicitly ask for review work -- please do not hesitate to ping explicitly.
- What is the final design you arrived at?
- What happens if a thread is unresponsive when shutdown happens, and later resumes OCaml work by taking the runtime/domain lock?

ghost commented:
> @dustanddreams I can see that you are rebasing this PR regularly, but I am a bit fuzzy about its current status.
> 
>     * Is it ready for review? My understanding is that you think it is, but are maybe to shy/polite to explicitly ask for review work -- please do not hesitate to ping explicitly.

Yes, it is ready for review. I am only rebasing it from time to time due to merge conflicts in `Changes`.

>     * What is the final design you arrived at?

Now that domain states are allocated dynamically, the logic is more careful to try not to deallocate locks or condition variables which are still in use. If there is only one domain left running at termination time, then we can proceed with the cleanup. Otherwise, we will `caml_log_gc` a warning and only cleanup what can be safely cleaned up (which is much less).

>     * What happens if a thread is unresponsive when shutdown happens, and later resumes OCaml work by taking the runtime/domain lock?

If a thread is unresponsive at shutdown time, it will get `pthread_cancel`ed by its backup thread in `stw_terminate_domain` (the STW rendezvous at shutdown, if memory cleanup is requested). But this fits in the "more than one domain still running at shutdown" so there will only be a minimal cleanup in this case.

kayceesrk commented on runtime/platform.c:
```diff
@@ -105,6 +105,24 @@ void caml_plat_mutex_free(caml_plat_mutex* m)
   check_err("mutex_free", pthread_mutex_destroy(m));
 }
 
+void caml_plat_mutex_reinit(caml_plat_mutex *m)
```
Is it correct to say that `caml_plat_mutex_reinit` must only be called on a mutex on which `caml_plat_mutex_init` had been called previously?

dustanddreams replied:
Yes, this is correct.

kayceesrk commented on runtime/platform.c:
```diff
@@ -105,6 +105,24 @@ void caml_plat_mutex_free(caml_plat_mutex* m)
+  } else {
+    /* lock was held at fork time, parent process still holds it, but we
+       don't and need to fix lock count */
+    DEBUG_UNLOCK(m);
```
Not specifically this PR, but what's the status of the lock here? The lock is held by the parent but not the current process. Is it safe to then call `caml_plat_mutex_init` on this locked mutex? https://pubs.opengroup.org/onlinepubs/007908799/xsh/pthread_mutex_init.html says,

> Attempting to initialise an already initialised mutex results in undefined behaviour.

How do I reason about the behaviour here?

dustanddreams replied:
After you `fork`, the state of the lock is unchanged. In the child process, which is a new process and obviously doesn't own this lock, we try to acquire it, which tells us whether the parent process holds the lock or not. This allows the debug counter of held locks to be adjusted correctly. It would be one unit too large if the parent process had the lock held at `fork` time.

dustanddreams replied:
Regarding the use of `caml_plat_mutex_init`, it's still an UB, and this PR does not change anything about that.

kayceesrk replied:
Ok

kayceesrk commented on runtime/caml/shared_heap.h:
```diff
@@ -105,6 +105,17 @@ void caml_cycle_heap(struct caml_heap_state*);
+/* Forces finalisation of all heap-allocated values,
+   disregarding both local and global roots.
+
+   Warning: finalisation is performed by means of forced sweeping, which may
```
What does "forced sweeping" mean?

dustanddreams replied:
This comment block has been moved from `runtime/caml/major_gc.h` and was added in `83093c4b2ee867e61b6f5692b0a6fcc18c38bc5a` ten years ago. I can't answer your question (-:

kayceesrk replied:
The comment before the warning is useful and descriptive enough to indicate that something terrible is being done to the heap state to finalise resources. I suggest we reword this comment to say, "Warning: This function should only be used on runtime shutdown".

dustanddreams replied:
Ok, done.

kayceesrk commented on runtime/domain.c:
```diff
@@ -938,7 +943,10 @@ void caml_update_minor_heap_max(uintnat requested_wsz) {
+  atomic_store_relaxed(&domains_exiting, 0);
+  atomic_store_relaxed(&caml_num_domains_running, 0);
+
+/* Use [caml_stat_calloc_noexc] to zero initialize [all_domains]. */
```
nit: indentation

dustanddreams replied:
Fixed.

kayceesrk commented on runtime/domain.c:
```diff
@@ -235,6 +237,7 @@ static atomic_uintnat /* dom_internal* */ stw_leader = 0;
 static dom_internal* all_domains;
+static atomic_intnat domains_exiting = 0;
 
 CAMLexport atomic_uintnat caml_num_domains_running;
```
```suggestion
CAMLexport atomic_uintnat caml_num_domains_running = 0;
```

dustanddreams replied:
Why not - although all global variables are zero-initialized anyway.

kayceesrk commented on runtime/domain.c:
```diff
@@ -1099,36 +1109,51 @@ static void install_backup_thread (dom_internal* di)
-    CAMLassert (msg == BT_INIT || /* Using fresh domain */
-                msg == BT_TERMINATE); /* Reusing domain */
-
+    if (msg != BT_TERMINATE)
```
When can this happen? We're trying to install a backup thread, and we see that the backup thread is running and the status is not `BT_TERMINATE`. How did we end up in this state? Was this backup thread left behind somehow?

dustanddreams replied:
I had to put this check at some point while working on this, but it turns out that it is apparently no longer necessary, and I can't trigger that test anymore, so I'll remove it.

kayceesrk commented on runtime/domain.c:
```diff
@@ -1099,36 +1109,51 @@ static void install_backup_thread (dom_internal* di)
-                msg == BT_TERMINATE); /* Reusing domain */
-
+    if (msg != BT_TERMINATE)
+      return;
```
Given that the function has a void return type, the caller has no way of knowing whether the installation of the backup thread succeeded. Is this a valid state to be in?

dustanddreams replied:
If spawning the backup thread fails, the function exits by raising an exception.

kayceesrk commented on runtime/domain.c:
```diff
@@ -1099,36 +1109,51 @@ static void install_backup_thread (dom_internal* di)
+{
+  CAMLassert(!caml_bt_is_self());
+
+  if (atomic_load_acquire(&di->backup_thread_running) != 0) {
```
```suggestion
  if (atomic_load_acquire(&di->backup_thread_running)) {
```

is more readable? (feel free to ignore).

kayceesrk commented on runtime/domain.c:
```diff
@@ -1278,6 +1303,10 @@ CAMLprim value caml_domain_spawn(value callback, value term_sync)
   pthread_t th;
   int err;
 
+  if (atomic_load_relaxed(&domains_exiting) != 0) {
```
```suggestion
  if (atomic_load_relaxed(&domains_exiting)) {
```
is more readable? (feel free to ignore).

kayceesrk commented on runtime/domain.c:
```diff
@@ -1321,7 +1350,9 @@ CAMLprim value caml_domain_spawn(value callback, value term_sync)
      ensure it is started here. */
-  install_backup_thread(domain_self);
+  domain_self->tid = pthread_self();
+  if (atomic_load_acquire(&domain_self->backup_thread_running) == 0)
```
Why is this guard necessary?

dustanddreams replied:
The backup thread for the main domain is spawned lazily: not until you spawn a second domain. This check makes sure `install_backup_thread` is only invoked once for the first domain, rather than on every domain spawn.

kayceesrk commented on runtime/domain.c:
```diff
@@ -2008,11 +2045,18 @@ static void domain_terminate (void)
     caml_orphan_finalisers(domain_state);
 
+    /* No need to check for interrupts if we are the last domain running. */
+    if (last)
```
If we do this, then we wouldn't emit `EV_DOMAIN_TERMINATE` event https://github.com/ocaml/ocaml/pull/12964/files#diff-67115925103982a8ebeb085cfab5ef31a182c9a442bc51e053934364d3750dafR2095.

dustanddreams replied:
Ah, yes indeed.

kayceesrk commented on runtime/domain.c:
```diff
@@ -2101,15 +2145,97 @@ static void domain_terminate (void)
      on caml_domain_alone (which uses caml_num_domains_running) in at least
      the shared_heap lockfree fast paths */
-  atomic_fetch_add(&caml_num_domains_running, -1);
+  if (!last)
```
Why is this guard needed?

dustanddreams replied:
Before this PR, `domain_terminate` is only invoked by domain threads, but never by the first domain. With this PR, the first domain invokes `domain_terminate` to perform more cleaning, but if it decrements the counter, then it would have a value of zero and `caml_domain_alone()` would return `false`. I guess I could change `caml_domain_alone` to use `<= 1` rather than `== 1`...

kayceesrk replied:
Either way, it needs a comment. I am leaning towards leaving the code as is and adding a comment rather than changing the implementation of `caml_domain_alone` where (a) we'll have to change the nice equality check with the inequality and (b) add a comment explaining why `caml_num_domains_running` could be `0`.

dustanddreams replied:
Done. I'll squash these last few commits into the earlier commits they belong to, once you're happy with the changes.

kayceesrk commented on runtime/startup_aux.c:
```diff
@@ -186,12 +190,21 @@ CAMLexport void caml_shutdown(void)
-  caml_finalise_heap();
+  if (!caml_domain_alone()) {
+    caml_gc_log("Some domains have not been joined prior to shutdown");
+    caml_stop_all_domains();
```
We try to stop all domains not just when the memory cleanup at exit mode is on. Have we decided that this is what we want?

dustanddreams replied:
This is `caml_shutdown`, which is only invoked in cleanup-at-exit mode, see `runtime/sys.c` lines 193-194.

kayceesrk commented on runtime/shared_heap.c:
```diff
@@ -603,6 +623,51 @@ intnat caml_sweep(struct caml_heap_state* local, intnat work) {
   return work;
 }
 
+/* Purgeing */
```
```suggestion
/* Purging */
```

kayceesrk reviewed (approved):
The PR looks mostly ok to me. Please look at my comments in `caml_domain_terminate`, where we break out early if we are the last domain. 

How did you verify that all the resources are being cleaned? Using `valgrind`? How do we ensure that the "memory cleanup at exit" is not accidentally broken by future change that leaks resources? Are there tests we can include such that the cleanup at exit mode is robust to future changes in the compiler?

ghost commented:
Rebased after applying a few suggestions.

kayceesrk commented:
There are failures in appveyor. See https://ci.appveyor.com/project/avsm/ocaml/builds/50717441.

kayceesrk commented:
@dustanddreams I'm interested in hearing your answer to the other questions raised in my comment at https://github.com/ocaml/ocaml/pull/12964#pullrequestreview-2341983314

ghost commented:
> How did you verify that all the resources are being cleaned? Using `valgrind`? How do we ensure that the "memory cleanup at exit" is not accidentally broken by future change that leaks resources? Are there tests we can include such that the cleanup at exit mode is robust to future changes in the compiler?

I have been testing with three configurations:
- regular `configure`, then check that all tests pass with `OCAMLRUNPARAM` empty or set to `OCAMLRUNPARAM=c=1,v=0`.
- `configure -fsanitize=leak`. You need to set `OCAMLRUNPARAM=c=1,v=0` prior to building, otherwise all invocations of `ocamlc` will report leaks. Once built, check that all tests pass, except for the few expected failures (such as `tests/embedded/cmcaml.ml` which has a real leak in `cmstub.c`). Also check with `USE_RUNTIME=d` while there.
- `configure --enable-tsan` and run the testsuite. This will cause a few false positives in tests which `fork()`, as the TSan runtime does not realize that locks held by the parent process are NOT held by the child process, and then complains about destroying locked mutexes during the cleanup.

`valgrind` can't really be used, as the use of separate stacks confuse it a lot.

To make sure the cleanup code gets exercized during tests and that the odds of it being broken without noticing is low, I suppose a few tests could be run with `OCAMLRUNPARAM` requiring cleanup. Or we could do something similar to TSan, add a CI build with `-fsanitize=leak` which would run the tests with `OCAMLRUNPARAM=c=1` and the few expected failures disabled. That run would only be done when changing the runtime (using a PR tag as done for TSan, or whatever). I am not sure there is an easier way to perform that kind of test, since it is not really possible to confirm the code works without building a leak-instrumented compiler.

kayceesrk commented:
Thanks for the explanation. 

Testing with `-fsanitize=leak` gives me the confidence that this works as intended. I do wonder whether it would be simplest to extend the current [extra (debug)](https://github.com/ocaml/ocaml/blob/e7dc11f290aa8d60cc95d576d42dced560dc691c/.github/workflows/build.yml#L115) run in the CI workflow to include leak sanitizer (while disabling the few expected failures)? 

Does the program run much slower with the leak sanitizer? The last CI debug run took 12m and there are a number of other workflows that took much longer (MSVC 64 bits took 24 minutes). It may be ok to stretch the current debug run a bit longer. What do you think? @gasche @dra27?

ghost commented:
> Does the program run much slower with the leak sanitizer? The last CI debug run took 12m and there are a number of other workflows that took much longer (MSVC 64 bits took 24 minutes). It may be ok to stretch the current debug run a bit longer. What do you think? @gasche @dra27?

Running the tests in sequential mode takes 50% more time with the leak sanitizer enabled on my machine (between 12 and 13 minutes instead of a bit more than 8 without).

kayceesrk commented:
Thanks. Given that leaks in the runtime may be caused by accidentally forgetting to free an allocated structure, I expect the issue to be caught fairly quickly running a test program. So I expect that running the entire testsuite will not be necessary. Perhaps the better idea is to modify the debug-s4096 workflow to add the leak check. It only runs the parallel tests, which should be sufficient.

kayceesrk commented:
The TSAN CI failure (https://github.com/ocaml/ocaml/actions/runs/11144860793/job/30973461200?pr=12964) has been observed elsewhere and shouldn't block this PR.

ghost commented:
~~I've tried to add some memleak tests in the github CI, but I probably don't know what I am doing.~~

kayceesrk commented:
@dustanddreams and I discussed on a different channel and decided that running the leak sanitizer tests in the CI should be considered in a different PR. I've made an issue to track this: https://github.com/ocaml/ocaml/issues/13515.

gasche commented:
(I don't have the bandwidth to look at this right now, so please feel free to make a decision without me if you can.)

kayceesrk commented:
@damiendoligez, IIUC, you have self-requested to review this PR. Do you plan to have a look?

kayceesrk commented:
@dustanddreams the failing `test_dropped_events` is due to a new data race introduced by this PR and will need to be investigated: https://github.com/ocaml/ocaml/actions/runs/11159629323/job/31018666175?pr=12964#step:5:575. I've removed the `merge-me` label for now. Note that this test is known to be flaky, but IIUC, the failure observed here, is not due to this flakiness. 

See related discussions at https://github.com/ocaml/ocaml/pull/13407#issuecomment-2381382979.

kayceesrk commented:
>  the failing test_dropped_events is due to a new data race introduced by this PR

Doesn't look like a new data race. I've proposed a fix in #13529.

ghost commented:
Thanks @kayceesrk for your work on the runtime event races.

kayceesrk commented on Changes:
```diff
@@ -40,6 +40,11 @@ Working version
+- #12964: Reintroduce "memory cleanup upon exit" mode. The cleanup will
+  however be incomplete if not all domains have been joined when the main
+  domain terminates.
+  (Miod Vallat, review by KC Sivaramakrishnan)
```
Please can you acknowledge @NickBarnes and @gasche, who reviewed parts of the PR and participated in the PR discussions?

dustanddreams replied:
Sure, done.

kayceesrk commented:
I believe that the PR is ready. The test failures in CI runs have been fixed in #13529. Since we haven't heard from @damiendoligez, I'll go ahead and merge. 

Thanks for the work and persevering through the various changes in the runtime @dustanddreams.
